core/src/banking_stage/transaction_scheduler/multi_iterator_consume_scheduler.rs

```diff
@@ -0,0 +1,617 @@
+        let pubkey = Pubkey::new_unique();
+        let mut container = create_container([
+            (&Keypair::new(), &[pubkey], 1, 1),
+            (&Keypair::new(), &[pubkey], 1, 2),
+        ]);
+
+        let num_scheduled = scheduler.schedule(&mut container).unwrap();
+        assert_eq!(num_scheduled, 2);
+        assert_eq!(
+            collect_work(&work_receivers[0]).1,
+            vec![txids!([1]), txids!([0])]
+        );
+    }
+
+    #[test]
+    fn test_schedule_consume_single_threaded_multi_batch() {
+        let (mut scheduler, work_receivers) = create_test_frame(1);
+        let mut container = create_container(
+            (0..4 * TARGET_NUM_TRANSACTIONS_PER_BATCH)
+                .map(|i| (Keypair::new(), [Pubkey::new_unique()], i as u64, 1)),
+        );
+
+        // expect 4 full batches to be scheduled
+        let num_scheduled = scheduler.schedule(&mut container).unwrap();
+        assert_eq!(num_scheduled, 4 * TARGET_NUM_TRANSACTIONS_PER_BATCH);
+
+        let thread0_work_counts: Vec<_> = work_receivers[0]
+            .try_iter()
+            .map(|work| work.ids.len())
+            .collect();
+        assert_eq!(thread0_work_counts, [TARGET_NUM_TRANSACTIONS_PER_BATCH; 4]);
+    }
+
+    #[test]
+    fn test_schedule_simple_thread_selection() {
+        let (mut scheduler, work_receivers) = create_test_frame(2);
+        let mut container =
+            create_container((0..4).map(|i| (Keypair::new(), [Pubkey::new_unique()], 1, i)));
+
+        let num_scheduled = scheduler.schedule(&mut container).unwrap();
+        assert_eq!(num_scheduled, 4);
+        assert_eq!(collect_work(&work_receivers[0]).1, [txids!([3, 1])]);
+        assert_eq!(collect_work(&work_receivers[1]).1, [txids!([2, 0])]);
+    }
+
+    #[test]
+    fn test_schedule_non_schedulable() {
+        let (mut scheduler, work_receivers) = create_test_frame(2);
+
+        let accounts = (0..4).map(|_| Keypair::new()).collect_vec();
+        let mut container = create_container([
+            (&accounts[0], &[accounts[1].pubkey()], 1, 2),
+            (&accounts[2], &[accounts[3].pubkey()], 1, 1),
+            (&accounts[1], &[accounts[2].pubkey()], 1, 0),
+        ]);
+
+        // high priority transactions [0, 1] do not conflict, and should be
+        // scheduled to *different* threads.
+        // low priority transaction [2] conflicts with both, and thus will
+        // not be schedulable until one of the previous transactions is
+        // completed.
+        let num_scheduled = scheduler.schedule(&mut container).unwrap();
+        assert_eq!(num_scheduled, 2);
+        let (thread_0_work, thread_0_ids) = collect_work(&work_receivers[0]);
+        assert_eq!(thread_0_ids, [txids!([0])]);
+        assert_eq!(collect_work(&work_receivers[1]).1, [txids!([1])]);
+
+        // Cannot schedule even on next pass because of lock conflicts
+        let num_scheduled = scheduler.schedule(&mut container).unwrap();
```

Comment:
Snapshotting the top-of-queue is an artifact of the MI approach, as we need a slice to iterate over.

Effectively, this is running multiple priority auctions within each slot - basically as often as the MI outer iteration(s) take. Where we allow new txs to enter the auction between outer iterations.

I agree streaming would probably be better. However, while we still require non-conflicting batches the MI approach is an efficient way to create these batches.

We still see significant performance benefit from batching, so I don't think we want to get rid of it entirely. See https://github.com/solana-labs/solana/pull/33332#discussion_r1337502718 - we also want to build these batches at the same time to distribute the work parallelly.

So let's look at what we could do instead of serializing the top-of-queue.
1. Pop - assign to thread 0
2. Pop - if conflict, move to some "blocked" buffer. Otherwise schedule to thread 1
3. Pop - if conflict, move to some "blocked" buffer. Otherwise schedule to thread 2
4. ...
Once our batches are full or we want to send them, we will need to push our "blocked" transactions back into our main queue (which may have been receiving while we were doing this). This may only push some subset of them in, because we can only unblock the ones that would be now (likely) be unblocked; however, it's still an O(nlogn) operation for every set of batches we send out. MI only pushes back into the main queue at the end of the outer iteration, each inner iteration is O(n).

Once we don't need to scan for non-conflicting txs, we might be able to take a streaming pop approach. As most conflicts would just be schedulable in the current batches - although we still need some sort of blocking buffer because conflicts across threads.